File: src/Disk.cpp
#include "../include/Disk.hpp"
Disk::Disk(int max_size){
    this -> max_size = max_size;
    qtt_occupied = 0;
    for(int i = 0; i < max_size;i++)
        memory.push_back("");
}
int Disk::FindValidSegment(int size){
    int cont = 0;
    for(unsigned int i = 0; i < memory.size();i++){
        if(i+size < memory.size()){
            if(memory[i] == "")
                for(unsigned int j = i; j < i+size;j++){
                    if(memory[j] == "")
                        cont++;
                    if(cont == size)
                        return i;
                }
        }
        else{
            break;
        }
        cont = 0;
    }
    return -1;
}
string Disk::CreateFile(string name, int initial_block, int size){
    if(max_size-qtt_occupied < size)
        return "Não pode criar o arquivo "+name+" (falta de espaço) .";
    else{
        if(initial_block > max_size-1 || initial_block < 0)
            return "Não pode criar o arquivo "+name+" (bloco inicial ocupado) .";
        else if(initial_block + size > max_size)
            return "Não pode criar o arquivo "+name+" (quantidade de blocos insuficientes no disco) .";
        else if(size < 0)
            return "Não pode criar o arquivo "+name+" (quantidade de blocos invalida, numero negativo) .";
        else if(initial_block < 0)
            return "Não pode criar o arquivo "+name+" (falta de espaço) .";
        else{
            vector<string> invalid;
            vector<string> valid;
            for(int i = initial_block;i < initial_block+size;i++){
                if(memory[i] != "")
                    invalid.push_back(to_string(i));
                else
                    valid.push_back(to_string(i));   
            }
            if(!invalid.empty()){
                if(invalid.size() == 1){
                    return "Não pode criar o arquivo "+name+" (bloco "+invalid[0]+" ocupado) .";
                }
                else{
                    string ret = "";
                    ret+= invalid[0];
                    for(unsigned i = 1; i<invalid.size()-1;i++){
                        ret+= ", "+invalid[i];
                    }
                    ret +=  " e " + invalid[invalid.size()-1];
                    return "Não pode criar o arquivo "+name+" (blocos "+ret+" ocupados) .";
                }
            }
            else{
                for(auto i : valid)
                    memory[stoi(i)] = name;
                if(valid.size() == 1)
                    return "Criou o arquivo "+name+" (bloco "+valid[0]+") .";
                else{
                    string ret = "";
                    ret+= valid[0];
                    for(unsigned i = 1; i<valid.size()-1;i++)
                        ret+= ", "+valid[i];
                    ret +=  " e " + valid[valid.size()-1];
                    return "Criou o arquivo "+name+" (blocos "+ret+") .";
                }
            }
        }
    }
}
string Disk::DeleteFile(string name){
    bool nameFound = false;
    for (unsigned int i = 0; i < memory.size();i++){
        if(memory[i] == name){
            memory[i] = "";
            nameFound = true;
        }
    }
    if(!nameFound)
        return "Não pode deletar o arquivo "+name+" porque ele não existe.";
    return "Deletou o arquivo X";
}

File: src/Disk.py
class Disk:

    def __init__(self, max_size):

        self.max_size = max_size
        self.memory = [("0",None)]*max_size
        self.qttOcuppied = 0

    def FindValidSegment(self,size):
        cont = 0
        for i in range(len(self.memory)):
            if (i+size <= len(self.memory)):
                if(self.memory[i][0] == "0"):
                    for j in range(i,i+size):
                        if(self.memory[j][0] == "0"):
                            cont += 1
                        if(cont == size):
                            return i
                else:
                    i += cont+1
            else:
                break
            cont = 0
        return -1


    def CreateFile(self,PID,process, name, initial_block, size):

        if(self.max_size-self.qttOcuppied < size):
            return "Nao pode criar o arquivo "+name+" (falta de espaco) ."
        elif(process == None):
            return "Nao pode criar o arquivo "+name+" (processo "+str(PID)+" nao existe)"
        else:
            if(initial_block >= self.max_size):
                return "Nao pode criar o arquivo "+name+" (bloco inicial ocupado) ."
            elif(initial_block + size >= self.max_size):
                return "Nao pode criar o arquivo "+name+" (quantidade de blocos insuficientes no disco) ."
            elif(size < 0):
                return "Nao pode criar o arquivo "+name+" (quantidade de blocos invalida, numero negativo) ."
            elif(initial_block < 0):
                return "Nao pode criar o arquivo "+name+" (falta de espaco) ."
            else:
                invalid = []
                valid = []
                for i in range(initial_block,initial_block+size):
                    if(self.memory[i][0] != "0"):
                        invalid.append(str(i))
                    else:
                        valid.append(str(i))   
                if(invalid):
                    if(len(invalid) == 1):
                        return "Nao pode criar o arquivo "+name+" (bloco "+invalid[0]+" ocupado) ."
                    
                    else:
                        ret = ""
                        if(invalid):
                            ret+= invalid[0]
                            for i in range(1,len(invalid)-1):
                                ret+= ", "+invalid[i]
                            
                            ret +=  " e " + invalid[len(invalid)-1]
                            return "Nao pode criar o arquivo "+name+" (blocos "+ret+" ocupados) ."
                    
                else:
                    for i in valid:
                        self.memory[int(i)] = (name,int(PID))
                        self.qttOcuppied += 1
                    if(len(valid) == 1):
                        return "Criou o arquivo "+name+" (bloco "+valid[0]+") ."
                    else:
                        ret = ""
                        ret+= valid[0]
                        for i in range(1,len(valid)-1):
                            ret+= ", "+valid[i]
                        ret +=  " e " + valid[len(valid)-1]
                        return "Criou o arquivo "+name+" (blocos "+ret+") ."
                
    def DeleteFile(self,PID,process, name):
        nameFound = False
        invalidProcess = False
        if(process == None):
            return "O processo "+str(PID)+" nao existe"
        
        for i in range(len(self.memory)):
            if(self.memory[i][0] == name):
                nameFound = True
                if(process["priority"] == 0 or self.memory[i][1] == process["PID"]):
                    self.memory[i] = "0"
                else:
                    invalidProcess = True
        msg = ""
        if(invalidProcess):
            msg = "O processo "+str(process["PID"])+" nao pode deletar o arquivo, pois ele nao tem permissao"
        if(not nameFound):
            if(msg != ""):
                msg += " "
            msg += "O processo "+str(process["PID"])+" nao pode deletar o arquivo "+name+" porque ele nao existe."
        if(invalidProcess or not nameFound):
            return msg        
        return "Deletou o arquivo "+name